crates/holochain/src/core/workflow/app_validation_workflow.rs

```diff
@@ -133,134 +133,129 @@ async fn app_validation_workflow_inner(
+                validate_op_outer(
+                    dna_hash.clone(),
+                    &op,
+                    &dht_op_hash,
+                    &conductor,
+                    &workspace,
+                    network,
+                    validation_dependencies.clone(),
+                )
+                .await
             }
-        }
-    });
-
-    let validation_results = futures::future::join_all(iter).await;
-
-    let sleuth_id = sleuth_id.clone();
-
-    let (accepted_ops, awaiting_ops, rejected_ops, activity) = workspace
-        .dht_db
-        .write_async(move |txn| {
-            let mut accepted = 0;
-            let mut awaiting = 0;
-            let mut rejected = 0;
-            let mut agent_activity = Vec::new();
-            for outcome in validation_results {
-                let (op_hash, dependency, op_lite, outcome, activity) = outcome;
-                // Get the outcome or return the error
-                let outcome = outcome.or_else(|outcome_or_err| outcome_or_err.try_into())?;
-
+            Err(e) => Err(e),
+        };
+        // Flatten nested app validation outcome to either ok or error
+        let validation_outcome = match validation_outcome {
+            Ok(outcome) => AppValidationResult::Ok(outcome),
+            Err(OutcomeOrError::Outcome(outcome)) => AppValidationResult::Ok(outcome),
+            Err(OutcomeOrError::Err(err)) => AppValidationResult::Err(err),
+        };
+
+        let sleuth_id = sleuth_id.clone();
+        match validation_outcome {
+            Ok(outcome) => {
                 // Collect all agent activity.
                 if let Some(activity) = activity {
                     // If the activity is accepted or rejected then it's ready to integrate.
                     if matches!(&outcome, Outcome::Accepted | Outcome::Rejected(_)) {
                         agent_activity.push(activity);
                     }
                 }
-
                 if let Outcome::AwaitingDeps(_) | Outcome::Rejected(_) = &outcome {
                     warn!(
                         msg = "DhtOp has failed app validation",
                         outcome = ?outcome,
                     );
                 }
-                match outcome {
-                    Outcome::Accepted => {
-                        accepted += 1;
-                        aitia::trace!(&hc_sleuth::Event::AppValidated {
-                            by: sleuth_id.clone(),
-                            op: op_hash.clone()
-                        });
-
-                        if dependency.is_none() {
-                            aitia::trace!(&hc_sleuth::Event::Integrated {
+
+                let accepted_ops = accepted_ops.clone();
+                let awaiting_ops = awaiting_ops.clone();
+                let rejected_ops = rejected_ops.clone();
+
+                let write_result = workspace
+                    .dht_db
+                    .write_async(move |txn| match outcome {
```

Comment:
The tradeoff is that each op is written to the database one by one.

Reply:
Probably not a huge problem because most of the time we're only validating a few ops at a time and not dealing with massive batches. I expect there will be lots of little things like this that we can tweak once we have regular performance testing in place (little bit of manifesting going on here for Wind Tunnel ;) )